cli/azd/pkg/containerapps/container_app.go

```diff
@@ -118,105 +161,151 @@ func (cas *containerAppService) AddRevision(
+	newRevisionResponse, err := revisionsClient.GetRevision(ctx, resourceGroupName, appName, newRevisionName, nil)
+	if err != nil {
+		return nil, fmt.Errorf("getting new revision '%s': %w", newRevisionName, err)
 	}
 
-	return nil
+	return &newRevisionResponse.Revision, nil
 }
 
-func (cas *containerAppService) syncSecrets(
+func (cas *containerAppService) ValidateRevision(
 	ctx context.Context,
 	subscriptionId string,
 	resourceGroupName string,
 	appName string,
-	containerApp *armappcontainers.ContainerApp,
-) (*armappcontainers.ContainerApp, error) {
-	// If the container app doesn't have any secrets, we don't need to do anything
-	if len(containerApp.Properties.Configuration.Secrets) == 0 {
-		return containerApp, nil
-	}
-
-	appClient, err := cas.createContainerAppsClient(ctx, subscriptionId)
+	revisionName string,
+) (*armappcontainers.Revision, error) {
+	revisionsClient, err := cas.createRevisionsClient(ctx, subscriptionId)
 	if err != nil {
 		return nil, err
 	}
 
-	// Copy the secret configuration from the current version
-	// Secret values are not returned by the API, so we need to get them separately
-	// to ensure the update call succeeds
-	secretsResponse, err := appClient.ListSecrets(ctx, resourceGroupName, appName, nil)
-	if err != nil {
-		return nil, fmt.Errorf("listing secrets: %w", err)
-	}
+	var result *armappcontainers.Revision
 
-	secrets := []*armappcontainers.Secret{}
-	for _, secret := range secretsResponse.SecretsCollection.Value {
-		secrets = append(secrets, &armappcontainers.Secret{
-			Name:  secret.Name,
-			Value: secret.Value,
-		})
-	}
+	// Wait 10 seconds between retries
+	backoff := retry.NewConstant(time.Second * 10)
 
-	containerApp.Properties.Configuration.Secrets = secrets
+	// Keep trying for up to 10 minutes or has reached a terminal state
+	maxDuration := retry.WithMaxDuration(time.Minute*10, backoff)
 
-	return containerApp, nil
+	err = retry.Do(ctx, maxDuration, func(ctx context.Context) error {
+		getRevisionResponse, err := revisionsClient.GetRevision(ctx, resourceGroupName, appName, revisionName, nil)
+		if err != nil {
+			return fmt.Errorf("getting revision '%s': %w", revisionName, err)
+		}
+
+		revision := getRevisionResponse.Revision
+
+		isProvisioned := false
+		isHealthy := false
+
+		switch *revision.Properties.ProvisioningState {
+		case armappcontainers.RevisionProvisioningStateProvisioning,
+			armappcontainers.RevisionProvisioningStateDeprovisioned,
+			armappcontainers.RevisionProvisioningStateDeprovisioning:
+			return retry.RetryableError(fmt.Errorf("revision '%s' is not ready", revisionName))
+		case armappcontainers.RevisionProvisioningStateFailed:
+			return fmt.Errorf("revision '%s' failed to provision, %s", revisionName, *revision.Properties.ProvisioningError)
+		case armappcontainers.RevisionProvisioningStateProvisioned:
+			isProvisioned = true
+		}
+
+		switch *revision.Properties.HealthState {
+		case armappcontainers.RevisionHealthStateUnhealthy:
```

Comment:
If an app is provisioned, but is unhealthy because the app is still warming up, and the health probes report Unhealthy, don't we end up failing here?

We should clarify what the design language of what "polling for complete app deployment" across host models is, and discuss if they should be configurable. We can start from AKS as the preliminary example, but I'd love to have these discussions.